### ISSUE TITLE
"limit number of pods that are allowed to run" failing on version-skewed clusters running 1.1 master and 1.0 nodes

### ISSUE THREAD

**Issue by ihmccreery** [kubernetes/kubernetes#15806]
See Jenkins job **kubernetes-upgrade-1.0-1.1-gke-step3-e2e-old**'s last few runs.

```
07:52:26   Validates MaxPods limit number of pods that are allowed to run. [It]
07:52:26   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/max_pods.go:158
07:52:26 
07:52:26   Expected
07:52:26       <[]api.Event | len:0, cap:0>: []
07:52:26   not to be empty
07:52:26 
07:52:26   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/max_pods.go:156
```

The relevant line is [max_pods.go:156](https://github.com/kubernetes/kubernetes/blob/release-1.0/test/e2e/max_pods.go#L156).  Note that this is running e2e tests off of the `release-1.0` branch.

It's possible that this is also caused by #15123, but it's not clear to me.

**Comment by ihmccreery** [comment]
This is also failing on kubernetes-upgrade-1.0-1.1-gke-step5-e2e-old, where we're running `release-1.0` tests against a cluster that's been fully upgraded, (master and nodes,) from 1.0 to 1.1.

**Comment by ihmccreery** [comment]
Status update: this may be caused by #15123, so that issue should be prioritized, but it's not totally clear if that will fix this problem.

**Comment by davidopp** [comment]
@gmarek said he would take a look and update the issue with any findings.

It may be a bit premature to look into it before #15123 is resolved, but at least we can try to figure out how to manually repro this now, so that we will know whether #15123 fixes this issue.

**Comment by gmarek** [comment]
I'm unable to find logs from failed runs, which makes everything slightly harder. I'll try to reproduce the problem on GCE cluster, but it may not work. FYI: most of `step5` was caused by waiting for skydns to come back up (which doesn't happen: #15227).

I see that there's a difference in number of system pods between version 1.1 and 1.0 - there's only 5 of them in 1.1 and 8 of them in 1.0. I'm wondering if this may cause problems in 'old' version of test which we're running there.

I'm going to try reproduce it on GCE (as this is the setup I understand better) and if I'll fail I'll try again on GKE.

**Comment by gmarek** [comment]
AFAICT there's a problem with the test, not the upgrade itself. The test from 1.1 branch fails on 1.0 cluster.

**Comment by gmarek** [comment]
It's caused by #12539. Can we please stop renaming things? AFAICT the problem is with test only, everything should be working 'on production'. In v1.0 scheduler Events reasons start with a small letter (e.g. failedScheduling) and v1.0 test is looking for those kind of events to occur. In v1.1 we changed Event reasons to start with a capital letter (e.g. FailedScheduling) and modified test accordingly. Effects are obvious.

We can:
- revert #12539
- patch it to v1.0
- merge and patch a test fix which adds handling both versions (#15948)

I'm for the last option, but whatever we do (except reverting) will break backward compatibility for everyone depending/counting/filtering/whatnot on scheduler Event reasons

**Comment by ihmccreery** [comment]
@gmarek thanks for your work on this.

From @bgrant0607's [comment](https://github.com/kubernetes/kubernetes/pull/12539/files#r36796328) on #12539:

> Do you have examples of systems interpreting the events? We've said that they shouldn't be treated as an API.

Furthermore, in my (rookie) opinion, #12539 seems like it should be considered a bug fix rather than an API change.

Thus I think #15948 makes good sense.

**Comment by davidopp** [comment]
Thanks a lot for debugging this, @gmarek, and I agree we should go with your PR.

I talked to Brian and he said we don't consider changes to values of fields in events to be breaking API changes. Unfortunately people do seem to depend on these in various places. One solution is what I proposed in #14541 so the compiler could prevent these incompatibilities (at least for components and tests that are compiled together).

**Comment by bgrant0607** [comment]
+1 to creating a file containing constants for the event reasons that are currently in use.

**Comment by ihmccreery** [comment]
I'll prepare a cherry pick.

**Comment by ihmccreery** [comment]
Reopening until cherry pick is merged.

**Comment by ihmccreery** [comment]
Cherry pick is #16186.

**Comment by davidopp** [comment]
The test mentioned in the first entry in this thread is still failing but not due to the error message related to this issue. Since the cherrypick has been merged, can we close this issue, and followup with additional debugging in #13602?

**Comment by ihmccreery** [comment]
Works for me.  Thanks to @gmarek again for your work on this.

**Comment by ihmccreery** [comment]
@davidopp I actually still see failures of the same kind in `kubernetes-upgrade-1.0-1.1-gke-step3-e2e-old` (master 1.1, nodes 1.0) and `kubernetes-upgrade-1.0-1.1-gke-step5-e2e-old` (master 1.1, nodes 1.1).  I assume this is because we haven't cherrypicked @gmarek's test fix into `release-1.0`.  So this is now a duplicate of #15116.
